spices/SPICE-0001-import-graph-analyzer-api.adoc

```diff
@@ -0,0 +1,336 @@
+  ["file:///path/to/my/subdir/foo.pkl"] {}
+}
+resolvedUris {
+  ["projectpackage://example.com/bar@1.0.0#/Bar.pkl"] = "file:///path/to/local/bar/Bar.pkl" // <2>
+  ["file:///foo/bar/Bar.pkl"] = "file:///foo/bar/bar.pkl" // <3>
+  ["package://example.com/bar@1.0.0#/Bar.pkl"] = "package://example.com/bar@1.0.0#/Bar.pkl" // <4>
+  ["pkl:reflect"] = "pkl:reflect"
+  ["pkl:jsonnet"] = "pkl:jsonnet"
+  ["pkl:xml"] = "pkl:xml"
+  ["pkl:protobuf"] = "pkl:protobuf"
+  ["file:///path/to/foo.pkl"] = "file:///path/to/foo.pkl"
+  ["file:///path/to/my/foo1.pkl"] = "file:///path/to/my/foo1.pkl"
+  ["file:///path/to/my/foo2.pkl"] = "file:///path/to/my/foo2.pkl"
+  ["file:///path/to/my/subdir/foo.pkl"] = "file:///path/to/my/subdir/foo.pkl"
+}
+----
+<1> dependency notation turns into `projectpackage://`.
+<2> local dependency's resolve URI is a local file.
+<3> module URIs that don't resolve to anything further stay the same.
+<4> package URIs stay the same.
+
+=== Gradle Plugin Changes
+
+The input files for various Pkl Gradle tasks will switch to using the Java `ImportAnalyzer` to build its input files.
+
+The set of input files will be `resolvedUris` whose keys start with `\file:///`.
+All other modules are assumed to be external to the machine and cache-able.
+
+The `transitiveModules` property is removed.
+
+== Future directions
+
+=== `pkl:anaylze` standard library module
+
+This is a starting point for the `pkl.analyze` stdlib module.
+Other forms of static analysis can be added to it as the need grows.
+
+=== Support in other language bindings
+
+This feature introduces `PklAnalyzer`, a new Java API that can analyze the imports of modules without requiring the evaluator.
+This is a capability that isn't being added to other language bindings.
+
+A simple workaround is to use the standard evaluator API, given the introduction of the new `pkl.anaylze` stdlib module, albeit with the drawback of the overhead of performing Pkl evaluation.
+
+To support evaluator-free analysis, a future improvement can be to add messages to the server API for performing import graph analysis.
+
+== Alternatives considered
+
+=== Enhance `pkl.reflect`
+
+One alternative is to enhance the `pkl.reflect` API to provide the import graph.
+
+Imagined API:
+
+[source,diff]
+----
+ module pkl.reflect
+
+ class Module {
++  /// The imports, extends, and amends declared by this module.
++  imports: Listing<Module>
+ }
+----
+
+The benefit of this approach is that an import can be defined in terms of regular Pkl evaluation.
+We do not need to provide a new CLI command, nor extension points in Gradle or server mode.
+
+Some drawbacks:
+
+* Reflection currently only provides information about a module's schema, using metadata that is already gathered when a module is initialized. Exposing this level of import requires that _data_ is also gathered during module initialization. This adds overhead to basic evaluation, even if the import graph is not needed.
```

Comment:
Import clauses/expressions are cheap to gather during parsing. What other data is required?

Reply:
I wouldn't call this "cheap"; adding runtime metadata is a new cost that we wouldn't have had to account for previously.

Reply:
The import clauses/expressions need to be parsed in any case. My assumption is that the only eager work remaining is to collect the parsed import URIs in a list.

Reply:
The cost here is storing metadata, rather than the runtime cost of parsing.

Reply:
More details would be required to understand the given justification. In general, pkl:reflect seems like a natural place for this feature.

Reply:
I dropped the thread here as we were busy working on other features. But, some more thoughts:

I looked into this further, and parsing is actually what takes the lions share of Pkl's time. It turns out that Pkl's parse is _quite_ slow.

However, there are some other reasons for why we'd like this here rather than in `pkl:reflect`:

* `pkl:reflect` requires that the module requires that the module be valid, which is a limitation that we don't necessarily want for this. In these cases, it's useful for an import analyzer to still be able to run. Examples:
** A module that circularly amends itself
** A module that declares invalid members (e.g. `hidden local`)
* the analyzer computes things that we probably don't want to pay for in `pkl:reflect`. For example, it expands glob imports, and also produces resolved URIs (which might involve fetching HTTP contents).

Reply:
I see. Fingers crossed that a hand-rolled parser will be faster than the ANTLR parser. (AFAIK, ANTLR parsers can be quite fast.)